DM-40002: Try to support pydantic v1 and v2

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Fix some tests that require optional dependencies

We always tend to install the optionals for testing but it
turns out the import tests were not working properly.

Consider removing S3 completely.

timj committed:
Fix types in test to match model

Somehow pydantic v1 did not notice.

timj committed:
Initial attempt to support pydantic v2

Still some failures. Lots of deprecation warnings.

timj committed:
For storage class conversion test use usable storage classes

AstropyArrow and Dataframe use classes that are not part of
the core butler dependencies. This leads to the tests failing
in a minimal environment. Replace with two storage classes
that are available to the core environment.

timj committed:
Remove fastapi pin

This enables pydantic v2

timj committed:
Use model_construct for pydantic v2

This is meant to be the rquivalent of using __setattr__
and __field_set__ directly. The latter has been renamed in
pydantic v2

timj committed:
Use mambaforge for build action

timj committed:
Use python 3.11 for doc build

timj committed:
Add python 3.10 definition of Self

timj committed:
Pydantic/fastAPI no longer handle generic Mapping for Config classes

This all worked before but now we have to force the Config
to a plain dict before it will be accepted.

timj committed:
Fix mypy with pydantic

There is clearly some issue with the RootModel constructor
that we will need to sort out.

timj committed:
Fix mypy for pydantic v1 and v2

This required that we allow unused igores since the location
of the ignores depends on the version of pydantic and mypy
always picks up the first definition it sees even if that
is a pydantic V2 code path and pydantic is v1.

andy-slac commented on tests/test_butler.py:
```diff
@@ -42,13 +42,14 @@
     boto3 = None
 
-    def mock_s3(cls):  # type: ignore[no-untyped-def]
+    def mock_s3(*args: Any, **kwargs: Any) -> Any:  # type: ignore[no-untyped-def]
```
I'd expect you can drop `type: ignore` here?

andy-slac commented on python/lsst/daf/butler/_compat.py:
```diff
@@ -0,0 +1,162 @@
+    Self = TypeVar("Self", bound="_BaseModelCompat")  # type: ignore
+
+
+PYDANTIC_V2 = PYDANTIC_VERSION.startswith("2.")
```
Pydantic 3.0 will be the same as 1.x? 🙂

timj replied:
I'm hoping that we've removed the v1 code by then...

andy-slac commented on python/lsst/daf/butler/_compat.py:
```diff
@@ -0,0 +1,162 @@
+            include: set[int] | set[str] | dict[int, Any] | dict[str, Any] | None = None,  # type: ignore
+            exclude: set[int] | set[str] | dict[int, Any] | dict[str, Any] | None = None,  # type: ignore
+            by_alias: bool = False,
+            skip_defaults: bool | None = None,
```
`skip_defaults` is completely unused?

timj replied:
Yes. `model_dump_json` does not have it. I can't really say what the difference is between `exclude_defaults` and `skip_defaults` and in fact its use has been deprecated even in pydantic v1:

```python
        if skip_defaults is not None:
            warnings.warn(
                f'{self.__class__.__name__}.json(): "skip_defaults" is deprecated and replaced by "exclude_unset"',
                DeprecationWarning,
            )
            exclude_unset = skip_defaults
```

andy-slac commented on python/lsst/daf/butler/_quantum_backed.py:
```diff
@@ -749,19 +745,40 @@ def _to_uuid_set(uuids: Iterable[str | uuid.UUID]) -> set[uuid.UUID]:
+                    key: SerializedDatastoreRecordData.direct(**records)
+                    for key, records in datastore_records.items()
+                },
+            )
```
Would it be possible to implement V1 `model_construct` to do the same thing as we do here to avoid if/else in every model class?

timj replied:
There is v1 `BaseModel.construct()` which does the same thing as `model_construct`. We need to understand why @natelust decided not to use it. If we know why and need to change we would need to go back to the explicit setters above and change to using the `__pydantic_fields_set__` property instead.

andy-slac reviewed (approved):
Looks OK after a quick look, one question is whether `_BaseModelCompat.model_construct` could be changed to avoid if/else in many model classes.

timj committed:
Add model_field and model_rebuild to v1 pydantic compat layer

timj committed:
Remove the --- from start of action yaml

timj commented:
@andy-slac you are right that I could implement v1 model_construct in _BaseModelCompat and just use it in the same way for both pydantic. Then I could even make v2 do the explicit setter approach if we decided we didn't want to use the base v2 base class model_construct. I will change.

timj committed:
Fix fallback model_fields class property

timj committed:
Use model_construct in direct methods

If the BaseModel.construct() method is slow we can still
reimplement it as was done in the individual direct() methods
but with the direct() methods still being much simpler.

timj committed:
Add note about use of BaseModel.construct

The previous version used by the direct() methods
would be something like:

        def model_construct(cls, _fields_set: set[str] | None = None, **values: Any) -> Self:
            node = cls.__new__(cls)
            setter = object.__setattr__
            for k, v in values.items():
                setter(node, k, v)
            if _fields_set is None:
                _fields_set = set(values.keys())
            setter(node, "__fields_set__", _fields_set)
            return node

Testing with 100MB graph indicates that construct() might be
fractionally slower than implementing it ourselves as above
but it is not conclusive. The direct() method only accounts
for 14% of the graph load time.

timj committed:
Add news fragment
